### DISCUSSION TITLE
Simulate inheritance

### DISCUSSION THREAD

**Discussion by Ghirigoro** [sinclairzx81/typebox#625] [Q&A] [answered]
Is there a way to simulate inheritance using Typebox? Say I have a type that looks like this:
```ts
type A = {
  id:string,
  elements: [...{type:string}[]]
}
```
And I need to create other types that extend it:
```ts
type B = {
  id:"B",
  elements: [{type:"B1", x:string}, {type:"B2", y:number}]
}
```
How would I create a schema for `B` that is constrained to the schema for `A`? Note that I also need the derived types be strongly typed (i.e. TS should be aware that the second element of `B` has a `y` field that is a `number`).

**Comment by sinclairzx81** [comment, marked as answer]
@Ghirigoro Hi,

You can use either Type.Intersect() or Type.Composite() to extend a base type. The example below extends using Composite.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4GUYCGMwAxnAL5wBmUEIcA5AAIDOwAdiQDYHBQD0MdBgBGEAB4MAsAChZJCOxbwAQgRYY4AXhTCAdAHkRAKwwkYACgSy4cYABMAXLsx78UDgHMLASiw24DC4MEAx2GBZnNFcAQSgoAlQLaIxDEzNLaxlbWyFMKP13L18A8h8fWTLZeUVlOAARDA8ANwx7bRdUgGE6SDYYDAsAbTUNHBS003MrAIcC1wAZYAGErgsGFQY-AKCQsIj51LiEpImAVXZgRWG4fn44M412ggfLxTgACybsOCgMMB4JE0AHdlh8UABXAGaYBUOCoCAQuABWwTIxTTI5LHYnG3e4YcSYcyBcQEcxcVBwGDAiCBYKhcIsPQonJ5DCHPRLFYENYbACMW382Wx4g5RXY3gqwtsZSFWLR6WmSBZuWEHK5TR56xUACZBSr4RyAHIQkAiJolaUUKW2AC65UqPnt1RkbLgo00OnwRFIAB42RA4R6AHyyN2NFptDre4gkf3CQMNJrAVr2YNAA)

```typescript
import { Type, Static } from '@sinclair/typebox'

const Base = Type.Object({
  id: Type.String(),
  elements: Type.Array(Type.Object({
    type: Type.String()
  }))
})

const Derived = Type.Composite([Base, Type.Object({
  id: Type.Literal('B'),
  elements: Type.Array(Type.Union([ // Used a Union here, replace with Tuple if you 
    Type.Object({                   // expect exactly two elements.
      type: Type.Literal('B1'),
      x: Type.String()
    }),
    Type.Object({ 
      type: Type.Literal('B2'),
      y: Type.Number()
    })
  ]))
})])

type Base = Static<typeof Base>
type Derived = Static<typeof Derived>
```
Above, the Derived type actually narrows the `id` and `elements` of the Base. Note there are other ways you could model this without extending (noting that TypeScript, and by extends TypeBox are structurally typed, so not nominal), but the above should work ok based on you types you provided.

Hope this helps
S
